radix-engine/src/engine/modules/execution_trace.rs

```diff
@@ -94,11 +291,266 @@ impl<R: FeeReserve> Module<R> for ExecutionTraceModule {
+        if !traced_input.is_empty() || !traced_output.is_empty() || !child_traces.is_empty() {
+            let trace = SysCallTrace {
+                origin,
+                sys_call_depth: self.current_sys_call_depth,
+                call_frame_actor: call_frame.actor.clone(),
+                call_frame_depth: call_frame.depth,
+                instruction_index,
+                input: traced_input,
+                output: traced_output,
+                children: child_traces,
+            };
+
+            let siblings = self
+                .sys_call_traces_stacks
+                .entry(self.current_sys_call_depth)
+                .or_insert(vec![]);
+            siblings.push(trace);
+        }
+
+        Ok(())
+    }
+
+    fn handle_processing_completed<R: FeeReserve>(
+        &mut self,
+        track: &mut Track<R>,
+    ) -> Result<(), ModuleError> {
+        // Some sanity checks
+        // No leftover inputs
+        if !self.traced_sys_call_inputs_stack.is_empty() {
+            return Err(ModuleError::ExecutionTraceError(
+                ExecutionTraceError::InvalidState("Leftover sys call inputs on stack".to_string()),
+            ));
+        }
+        // At most one entry in call traces mapping (the root level)
+        if self.sys_call_traces_stacks.len() > 1 {
+            return Err(ModuleError::ExecutionTraceError(
+                ExecutionTraceError::InvalidState("Leftover sys call traces on stack".to_string()),
+            ));
+        }
+
+        for (_, traces) in self.sys_call_traces_stacks.drain() {
+            // Emit an output event for each "root" sys call trace
+            for trace in traces {
+                track.add_tracked_event(TrackedEvent::SysCallTrace(trace));
+            }
+        }
+
+        Ok(())
+    }
+
+    fn extract_trace_data(
+        heap: &mut Heap,
+        traceable: &dyn Traceable,
+    ) -> Result<TracedSysCallData, ModuleError> {
+        let mut buckets: HashMap<BucketId, Resource> = HashMap::new();
+        for bucket_id in traceable.buckets() {
+            let bucket_resource = Self::read_bucket_resource(heap, &bucket_id)?;
+            buckets.insert(bucket_id, bucket_resource);
+        }
+
+        let mut proofs: HashMap<ProofId, ProofSnapshot> = HashMap::new();
+        for proof_id in traceable.proofs() {
+            let proof = Self::read_proof(heap, &proof_id)?;
+            proofs.insert(proof_id, proof);
+        }
+
+        Ok(TracedSysCallData { buckets, proofs })
+    }
+
+    fn read_proof(heap: &mut Heap, proof_id: &ProofId) -> Result<ProofSnapshot, ModuleError> {
+        let node_id = RENodeId::Proof(proof_id.clone());
+        let substate_ref = heap
+            .get_substate(node_id, &SubstateOffset::Proof(ProofOffset::Proof))
+            .map_err(|e| {
+                ModuleError::ExecutionTraceError(ExecutionTraceError::CallFrameError(e))
+            })?;
+        Ok(substate_ref.proof().into())
```

Comment:
Ah, missed this `into`. Suggeset using associated method `.snapshot` instead of `into` trait.

Into is for
> A value-to-value conversion that consumes the input value. 

`.into()` gives false impression that the proof is consumed (but in fact it's the reference that gets consumed).

Reply:
Ahh, right. Fixed